lib/rex/post/sql/ui/console/command_dispatcher/client.rb

```diff
@@ -0,0 +1,168 @@
+          #
+          # Initializes an instance of the core command set using the supplied console
+          # for interactivity.
+          #
+          # @param console The protocol-specific Rex Ui Console
+          def initialize(console)
+            super
+
+            @db_search_results = []
+          end
+
+          #
+          # List of supported commands.
+          #
+          def commands
+            cmds = {
+              'query'   => 'Run a raw SQL query',
+              'shell'   => 'Enter a raw shell where SQL queries can be executed',
+            }
+
+            reqs = {}
+
+            filter_commands(cmds, reqs)
+          end
+
+          def name
+            raise ::NotImplementedError
+          end
+
+          def help_args?(args)
+            return false unless args.instance_of?(::Array)
+
+            args.include?('-h') || args.include?('--help')
+          end
+
+          def cmd_shell_help
+            print_line 'Usage: shell'
+            print_line
+            print_line 'Go into a raw SQL shell where SQL queries can be executed.'
+            print_line "To exit, type 'exit', 'quit', 'end' or 'stop'."
+            print_line
+          end
+
+          def cmd_shell(*args)
+            if help_args?(args)
+              cmd_shell_help
+              return
+            end
+
+            console = shell
+            # Pass in self so that we can call cmd_query in subsequent calls
+            console.interact_with_client(client_dispatcher: self)
+          end
+
+          def normalise_sql_result(result)
+            # Let's extract the columns and rows as that's the only thing we want?
+            # We can special-case it for each result type/class.
+            normalised_sql = { columns: [], rows: [], errors: [] }
+
+            case result
+            when ::Mysql::Result # MySQL
+              normalised_sql[:columns] = result.fields.each.map(&:name)
+              normalised_sql[:rows] = result.entries
+              # MySQL errors are handled by raising an exception when querying,
+              # meaning we don't have that in the Result object and instead we use the 'when :Hash' branch
+              normalised_sql[:errors] = []
+            when ::Msf::Db::PostgresPR::Connection::Result # PostgreSQL
+              normalised_sql[:columns] = result.fields.each.map(&:name)
+              normalised_sql[:rows] = result.rows
+              # PostgreSQL errors are handled by raising an exception when querying,
+              # meaning we don't have that in the Result object and instead we use the 'when :Hash' branch
+              normalised_sql[:errors] = []
+            when ::Hash # MSSQL
+              normalised_sql[:columns] = result[:colnames]
+              normalised_sql[:rows] = result[:rows]
+              normalised_sql[:errors] = result[:errors]
+            else
+              raise 'Unknown SQL query result type'
+            end
```

Comment:
Rather than special casing for each session type can we just implement `normalise_sql_result` in the sessions `command_dispatcher/client.rb`